pkg/kubelet/server/stats/summary.go

```diff
@@ -0,0 +1,191 @@
+
+	rootInfo, found := infos["/"]
+	if !found || len(rootInfo.Stats) < 1 || latestContainerStats(&rootInfo) == nil {
+		return nil, fmt.Errorf("Missing stats for root container")
+	}
+
+	rootStats := containerInfoV2ToStats("", &rootInfo)
+	nodeStats := NodeStats{
+		NodeName: node.Name,
+		CPU:      rootStats.CPU,
+		Memory:   rootStats.Memory,
+		Network:  containerInfoV2ToNetworkStats(&rootInfo),
+	}
+
+	systemContainers := map[string]string{
+		SystemContainerKubelet: nodeConfig.KubeletContainerName,
+		SystemContainerRuntime: nodeConfig.DockerDaemonContainerName, // TODO: add support for other runtimes
+		SystemContainerMisc:    nodeConfig.SystemContainerName,
+	}
+	for sys, name := range systemContainers {
+		if info, ok := infos[name]; ok {
+			nodeStats.SystemContainers = append(nodeStats.SystemContainers, containerInfoV2ToStats(sys, &info))
+		}
+	}
+
+	summary := Summary{
+		Time: unversioned.NewTime(latestContainerStats(&rootInfo).Timestamp),
+		Node: nodeStats,
+		Pods: buildSummaryPods(infos),
+	}
+	return &summary, nil
+}
+
+func latestContainerStats(info *cadvisorapiv2.ContainerInfo) *cadvisorapiv2.ContainerStats {
+	stats := info.Stats
+	return stats[len(stats)-1]
+}
+
+// buildSummaryPods aggregates and returns the container stats in cinfos by the Pod managing the container.
+// Containers not managed by a Pod are omitted.
+func buildSummaryPods(cinfos map[string]cadvisorapiv2.ContainerInfo) []PodStats {
+	// Map each container to a pod and update the PodStats with container data
+	podToStats := map[NonLocalObjectReference]*PodStats{}
+	for _, cinfo := range cinfos {
+		// Build the Pod key if this container is managed by a Pod
+		if !isPodManagedContainer(&cinfo) {
+			continue
+		}
+		ref := buildPodRef(&cinfo)
+
+		// Lookup the PodStats for the pod using the PodRef.  If none exists, initialize a new entry.
+		stats, found := podToStats[ref]
+		if !found {
+			stats = &PodStats{PodRef: ref}
+			podToStats[ref] = stats
+		}
+
+		// Update the PodStats entry with the stats from the container by adding it to stats.Containers
+		containerName := dockertools.GetContainerName(&cinfo)
+		if containerName == leaky.PodInfraContainerName {
+			// Special case for infrastructure container which is hidden from the user and has network stats
+			stats.Network = containerInfoV2ToNetworkStats(&cinfo)
+		} else {
+			stats.Containers = append(stats.Containers, containerInfoV2ToStats(containerName, &cinfo))
+		}
+	}
+
+	// Add each PodStats to the result
+	result := make([]PodStats, 0, len(podToStats))
+	for _, stats := range podToStats {
+		result = append(result, *stats)
+	}
+	return result
+}
+
+// buildPodRef returns a NonLocalObjectReference that identifies the Pod managing cinfo
+func buildPodRef(cinfo *cadvisorapiv2.ContainerInfo) NonLocalObjectReference {
+	// TODO: Move these out of dockertools and into a general container pkg + make sure rkt is supported
+	podName := dockertools.GetPodName(cinfo)
```

Comment:
We should go through the `container.Runtime` interface here. Otherwise it is one more magic gotcha for rocket integration.